examples/pytorch/schnet/layers.py

```diff
@@ -0,0 +1,302 @@
+        return {"new_node": new_node}
+
+    def forward(self, g):
+        g.apply_edges(self.update_edge)
+        g.send_and_recv(g.edges(),
+                        message_func=self.send_edge_info,
+                        reduce_func=self.elemnet_wise_gather)
+        return g.ndata["new_node"]
+
+
+class Interaction(nn.Module):
+    """
+    The interaction layer in the SchNet model.
+    """
+
+    def __init__(self, rbf_dim, dim):
+        super().__init__()
+        self._node_dim = dim
+        self.activation = nn.Softplus(beta=0.5, threshold=14)
+        self.node_layer1 = nn.Linear(dim, dim, bias=False)
+        self.cfconv = CFConv(rbf_dim, dim, act=self.activation)
+        self.node_layer2 = nn.Linear(dim, dim)
+        self.node_layer3 = nn.Linear(dim, dim)
+
+    def forward(self, g):
+
+        g.ndata["new_node"] = self.node_layer1(g.ndata["node"])
+        cf_node = self.cfconv(g)
+        cf_node_1 = self.node_layer2(cf_node)
+        cf_node_1a = self.activation(cf_node_1)
+        new_node = self.node_layer3(cf_node_1a)
+        g.ndata["node"] = g.ndata["node"] + new_node
+        return g.ndata["node"]
+
+
+class VEConv(nn.Module):
+    """
+    The Vertex-Edge convolution layer in MGCN which take edge & vertex features
+    in consideratoin at the same time.
+    """
+
+    def __init__(self, rbf_dim, dim=64, update_edge=True):
+        """
+        Args:
+            dim: the dimension of linear layers
+            act: activation function (default shifted softplus)
+        """
+        super().__init__()
+        self._rbf_dim = rbf_dim
+        self._dim = dim
+        self._update_edge = update_edge
+
+        self.linear_layer1 = nn.Linear(self._rbf_dim, self._dim)
+        self.linear_layer2 = nn.Linear(self._dim, self._dim)
+        self.linear_layer3 = nn.Linear(self._dim, self._dim)
+
+        self.activation = nn.Softplus(beta=0.5, threshold=14)
+
+    def update_rbf(self, edges):
+        rbf = edges.data["rbf"]
+        h = self.linear_layer1(rbf)
+        h = self.activation(h)
+        h = self.linear_layer2(h)
+        return {"h": h}
+
+    def update_edge(self, edges):
+        edge_f = edges.data["edge_f"]
+        h = self.linear_layer3(edge_f)
+        return {"edge_f": h}
+
+    def send_edge_info(self, edges):
+        update_info = edges.src["new_node"] * edges.data["h"] + edges.data[
+            "edge_f"]
+        return {"neighbor_info": update_info}
+
+    def elemnet_wise_gather(self, nodes):
+        """
+        Implement element-wise multiplication (node x edge)
+        and reduce (sum) to nodes.
+        """
+        new_node = th.sum(nodes.mailbox["neighbor_info"], dim=1)
+        return {"new_node": new_node}
+
+    def forward(self, g):
+        g.apply_edges(self.update_rbf)
+        if self._update_edge:
+            g.apply_edges(self.update_edge)
+
+        g.send_and_recv(g.edges(),
+                        message_func=self.send_edge_info,
```

Comment:
The same suggestion as above, your `send_edge_info` could be decomposed to a `u_mul_e` and a `copy_e`:
```suggestion
                        message_func=[fn.u_mul_e('new_node', 'h', 'm_0'), fn.copy_e('edge_f', 'm_f', 'm_1)],
                        reduce_func=[fn.sum('m_0', 'new_node_0'), fn.sum('m_1', 'new_node_1')])
            g.ndata['new_node'] = g.ndata.pop('new_node_0') + g.ndata.pop('new_node_1')
```